src/iceberg/expression/aggregate.h

```diff
@@ -0,0 +1,230 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#pragma once
+
+/// \file iceberg/expression/aggregate.h
+/// Aggregate expression definitions.
+
+#include <memory>
+#include <optional>
+#include <string>
+#include <vector>
+
+#include "iceberg/expression/expression.h"
+#include "iceberg/expression/term.h"
+#include "iceberg/result.h"
+#include "iceberg/type_fwd.h"
+
+namespace iceberg {
+
+/// \brief Base aggregate holding an operation and a term.
+template <TermType T>
+class ICEBERG_EXPORT Aggregate : public virtual Expression {
+ public:
+  ~Aggregate() override = default;
+
+  Expression::Operation op() const override { return operation_; }
+
+  const std::shared_ptr<T>& term() const { return term_; }
+
+ protected:
+  Aggregate(Expression::Operation op, std::shared_ptr<T> term)
+      : operation_(op), term_(std::move(term)) {}
+
+  Expression::Operation operation_;
+  std::shared_ptr<T> term_;
+};
+
+/// \brief Base class for unbound aggregates.
+class ICEBERG_EXPORT UnboundAggregate : public virtual Expression,
+                                        public Unbound<Expression> {
+ public:
+  ~UnboundAggregate() override = default;
+
+  bool is_unbound_aggregate() const override { return true; }
+};
+
+/// \brief Template for unbound aggregates that carry a term and operation.
+template <typename B>
+class ICEBERG_EXPORT UnboundAggregateImpl : public UnboundAggregate,
+                                            public Aggregate<UnboundTerm<B>> {
+  using BASE = Aggregate<UnboundTerm<B>>;
+
+ public:
+  static Result<std::shared_ptr<UnboundAggregateImpl<B>>> Make(
+      Expression::Operation op, std::shared_ptr<UnboundTerm<B>> term);
+
+  std::shared_ptr<NamedReference> reference() override {
+    return BASE::term() ? BASE::term()->reference() : nullptr;
+  }
+
+  Result<std::shared_ptr<Expression>> Bind(const Schema& schema,
+                                           bool case_sensitive) const override;
+
+  std::string ToString() const override;
+
+ private:
+  static constexpr bool IsSupportedOp(Expression::Operation op) {
+    return op == Expression::Operation::kCount ||
+           op == Expression::Operation::kCountNull ||
+           op == Expression::Operation::kCountStar || op == Expression::Operation::kMax ||
+           op == Expression::Operation::kMin;
+  }
+
+  UnboundAggregateImpl(Expression::Operation op, std::shared_ptr<UnboundTerm<B>> term)
+      : BASE(op, std::move(term)) {
+    ICEBERG_DCHECK(IsSupportedOp(op), "Unexpected aggregate operation");
+    ICEBERG_DCHECK(op == Expression::Operation::kCountStar || BASE::term() != nullptr,
+                   "Aggregate term cannot be null unless COUNT(*)");
+  }
+};
+
+/// \brief Base class for bound aggregates.
+class ICEBERG_EXPORT BoundAggregate : public Aggregate<BoundTerm>, public Bound {
+ public:
+  using Aggregate<BoundTerm>::op;
+  using Aggregate<BoundTerm>::term;
+
+  class ICEBERG_EXPORT Aggregator {
```

Comment:
```suggestion
  class Aggregator {
```

I think we don't need to add export for inner class.